01_roman_numerals.py

```diff
@@ -0,0 +1,191 @@
+  """
+  I - 1
+  V - 5
+  X - 10
+  L - 50
+  C - 100
+  D - 500
+  M - 1000
+  """
+
+  def roman_to_number(self, string):
+    """
+    Converts a string to roman numerals.
+    Assumes that everything in the string is a valid roman numeral
+    Subtraction is not present. ie: IX = 9 does not exist. 9 would be XIIII
+
+    Analysis
+    Time - o(n)
+    * have to iterate through the entire string once. n is the length of the string
+    Space - o(1)
+    * size of lookup table is constant since there is a fixed set of roman numerals
+    * rest are variables that are integers
+    """
+    lookup = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
+
+    acc = 0
+
+    for c in string:
+      acc += lookup[c]
+
+    return acc
+
+  def roman_to_number_2(self, string):
+    """
+      Converts a string to roman numerals.
+      Assumes that everything in the string is a valid roman numeral that follows the rules below.
+      Subtraction is present. ie: IX = -1+10 = 9
+      Numbers go bigger to smaller unless subtraction is present.
+      Not all pairs are allowed.
+      * I can precede V or X
+      * X can precede L or C
+      * C can precede D or M
+
+      Analysis
+      Time - o(n) where n is the length of the string
+      * no matter what, we have to iterate through each character in the
+      entire string once
+      Space - o(1)
+      * lookup table is a constant
+      * temporary variables store either fixed sized strings or integers
+    """
+    lookup = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
+
+    acc = 0
+    idx = 0
+    str_length = len(string)
+
+    while idx < str_length:
+      curr_num_in_integer = lookup[string[idx]]
+      if ((idx + 1) < str_length):
+        next_num_in_integer = lookup[string[idx+1]]
+        if next_num_in_integer > curr_num_in_integer:
+          acc += (next_num_in_integer - curr_num_in_integer)
+          idx += 2
+        else:
+          acc += curr_num_in_integer
+          idx += 1
+      else:
+        acc += curr_num_in_integer
+        idx += 1
+
+    return acc
+
+  def roman_to_number_3(self, string):
+    """
+    Converts a string to roman numerals.
+    Assumes that everything in the string is a valid roman numeral
+    that follows the rules below.
+    Subtraction is present. ie: IX = -1+10 = 9
+    Numbers go bigger to smaller unless subtraction is present.
+    Not all pairs are allowed.
+    * I can precede V or X
+    * X can precede L or C
+    * C can precede D or M
+
+    Inspired by Elliott Jin's suggestion to use the lookup table from the
+    original solution for roman_to_number/2
+
+    Analysis
+    time - o(nk)?
+    * not very sure. lookup length is constant but we have to go through the whole string as well
+    * so maybe o(n) where n is length of the string?
+    * each iteration uses o(k) time to slice the string in `string[len(elem[0]):]`
+    * also not sure how long `string.startswith` takes
```

Comment:
One way I estimate time complexity for library functions is just to imagine how I'd implement something.  The library function should be at least that efficient!

For example, one easy way to implement `startswith` is to check each character one by one.  So if the string has length `n` and the thing you're checking has length `m`, the time complexity is likely O(`min(n, m)`).